dsgrid/dataset/pivoted_table.py

```diff
@@ -0,0 +1,197 @@
+                        f"Mapping dimension query name {query_name} produced duplicate to_ids for one or more from_ids"
+                    )
+            from_fractions = get_unique_values(records, "from_fraction")
+            if len(from_fractions) != 1 and float(next(iter(from_fractions))) != 1.0:
+                # TODO: This needs to apply from_fraction to each load value column.
+                # Also needs to handle all possible from_id/to_id combinationall possible from_id/to_id combinations
+                # If aggregation is not allowed then it should raise an error.
+                raise DSGInvalidParameter(
+                    f"Mapping dimension query name {query_name} produced from_fractions other than 1.0: {from_fractions}"
+                )
+            else:
+                records = records.drop("from_fraction")
+                if column.function is not None:
+                    # TODO: Do we want to allow this?
+                    raise Exception(
+                        f"Applying a SQL function to added column={query_name} is not supported yet"
+                    )
+                df = (
+                    df.join(records, on=df[base_query_name] == records.from_id)
+                    .drop("from_id")
+                    .withColumnRenamed("to_id", query_name)
+                )
+                context.add_dimension_query_name(
+                    supp_dim.model.dimension_type, query_name, dataset_id=self.dataset_id
+                )
+
+        return df
+
+    def process_aggregations(
+        self, df, aggregations: List[AggregationModel], context: QueryContext
+    ):
+        df = self.process_pivoted_aggregations(df, aggregations, context)
+        return self.process_stacked_aggregations(df, aggregations, context)
+
+    def process_pivoted_aggregations(
+        self, df, aggregations: List[AggregationModel], context: QueryContext
+    ):
+        """Aggregate the pivoted dimensional data as specified by aggregations.
+
+        Parameters
+        ----------
+        df : pyspark.sql.DataFrame
+        aggregations : AggregationModel
+        context : QueryContext
+
+        Returns
+        -------
+        pyspark.sql.DataFrame
+
+        """
+        if not aggregations:
+            return df
+
+        pivoted_columns = copy.deepcopy(context.get_pivoted_columns(dataset_id=self.dataset_id))
+        new_pivoted_columns = []  # Will be the pivoted columns after the last aggregation.
+        pivoted_dim_type = context.get_pivoted_dimension_type(dataset_id=self.dataset_id)
+        for agg in aggregations:
+            base_query_names = self.project_config.get_base_dimension_query_names()
+            dimension_query_name = None
+            for _, column in agg.iter_dimensions_to_keep():
+                query_name = column.dimension_query_name
+                # No work is required if the user requested the base dimension for the pivoted
+                # dimension.
+                if (
```

Comment:
Is this cond checking if query_name is supplemental?

Reply:
I feel like we can refactor these checks to return whether a query_name is base vs. supplemental, and stacked vs. pivoted.

Reply:
Idea:

Create a new class to store dimension query name information. Store this for every query name in the context. Add accessor methods. Add methods like `should_process_aggregation(query_name)`.
```
{
    "query_name1": {"dimension_type": "base", "format_type": "pivoted"},
     "query_name2": {"dimension_type": "supplemental", "format_type": "stacked"}
}
```

Reply:
@lixiliu In the best case we can track whether it's a base or supplemental dimension. We are already tracking pivoted columns per dataset and for the project (they could be different across datasets - and the project might want to change to long format). We already have a mechanism to identify base vs supplemental from the project config. Maybe that could be simplified...but not by much.